src/goto-instrument/cover_basic_blocks.cpp

```diff
@@ -0,0 +1,227 @@
+        if(
+          in_t->is_goto() && !in_t->is_backwards_goto() &&
+          in_t->guard.is_true())
+        {
+          current_block = block_map[in_t];
+          increase_block_nr = false;
+        }
+      }
+      if(increase_block_nr)
+      {
+        block_infos.push_back(block_infot());
+        block_infos.back().representative_inst = it;
+        block_infos.back().source_location = source_locationt::nil();
+        current_block = block_infos.size() - 1;
+      }
+    }
+
+    INVARIANT(
+      current_block < block_infos.size(), "current block number out of range");
+    block_infot &block_info = block_infos.at(current_block);
+
+    block_map[it] = current_block;
+
+    // update lines belonging to block
+    const irep_idt &line = it->source_location.get_line();
+    if(!line.empty())
+      block_info.lines.insert(unsafe_string2unsigned(id2string(line)));
+
+    // set representative program location to instrument
+    if(
+      !it->source_location.is_nil() &&
+      !it->source_location.get_file().empty() &&
+      !it->source_location.get_line().empty() &&
+      block_info.source_location.is_nil())
+    {
+      block_info.representative_inst = it; // update
+      block_info.source_location = it->source_location;
+    }
+
+    next_is_target =
+#if 0
+      // Disabled for being too messy
+      it->is_goto() || it->is_function_call() || it->is_assume();
+#else
+      it->is_goto() || it->is_function_call();
+#endif
+  }
+
+  for(auto &block_info : block_infos)
+    update_covered_lines(block_info);
+}
+
+unsigned cover_basic_blockst::block_of(goto_programt::const_targett t) const
+{
+  block_mapt::const_iterator it = block_map.find(t);
+  INVARIANT(it != block_map.end(), "instruction must be part of a block");
+  return it->second;
+}
+
+goto_programt::const_targett
+cover_basic_blockst::instruction_of(unsigned block_nr) const
+{
+  INVARIANT(block_nr < block_infos.size(), "block number out of range");
+  return block_infos.at(block_nr).representative_inst;
+}
+
+const source_locationt &
+cover_basic_blockst::source_location_of(unsigned block_nr) const
+{
+  INVARIANT(block_nr < block_infos.size(), "block number out of range");
+  return block_infos.at(block_nr).source_location;
+}
+
+void cover_basic_blockst::select_unique_java_bytecode_indices(
+  const goto_programt &goto_program,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+  std::set<unsigned> blocks_seen;
+  std::set<irep_idt> bytecode_indices_seen;
+
+  forall_goto_program_instructions(it, goto_program)
+  {
+    unsigned block_nr = block_of(it);
+    if(blocks_seen.find(block_nr) != blocks_seen.end())
+      continue;
+
+    INVARIANT(block_nr < block_infos.size(), "block number out of range");
+    block_infot &block_info = block_infos.at(block_nr);
+    if(block_info.representative_inst == goto_program.instructions.end())
+    {
+      if(!it->source_location.get_java_bytecode_index().empty())
+      {
```

Comment:
This seems very language specific; what happens if the input is not JVM code?  Would it be better to separate basic blocks and JVM instructions as different coverage options?

Reply:
Yes, that's a good idea. Noted for a future refactoring: https://github.com/diffblue/cbmc/issues/1686